our assert() is not compatible with the C one, and breaks in edge cases

Reported here as an incompatibility with perl 5.36: https://cyrus.topicbox.com/groups/info/T4dbb95a05bf1581e-Mdc484c415098530c67b67f26

We need to either redefine our `assert()` in a way that's compatible with the C one (losing the `{}` and the `if()`), or stop trying to use it as a general replacement for the C one.

> so the right thing to do is probably rename the Cyrus one.

The problem that I saw when testing with perl 5.36 is that, for example, `managesieve.xs` includes `perl.h`, which itself includes `<assert.h>`. The include path has the cyrus `lib/` early, and as a result the _wrong_ assert.h is pulled in, mixing OS and cyrus headers.

```
In file included from /usr/perl5/5.36/lib/i86pc-solaris-thread-multi-64/CORE/perl.h:3855,
                 from managesieve.xs:46:
/usr/perl5/5.36/lib/i86pc-solaris-thread-multi-64/CORE/sv_inline.h: In function 'Perl_newSV_type':
../../../lib/assert.h:47:25: error: expected expression before '{' token
   47 | #define assert(ex)      {if (!(ex))assertionfailed(__FILE__, __LINE__, #ex);}
      |                         ^
/usr/perl5/5.36/lib/i86pc-solaris-thread-multi-64/CORE/handy.h:2787:28: note: in expansion of macro 'assert'
 2787 | #define perl_assert_ptr(p) assert( ((void*)(p)) != 0 )
      |                            ^~~~~~
/usr/perl5/5.36/lib/i86pc-solaris-thread-multi-64/CORE/handy.h:2792:47: note: in expansion of macro 'perl_assert_ptr'
 2792 | #define Zero(d,n,t)     (MEM_WRAP_CHECK_(n,t) perl_assert_ptr(d), (void)memzero((char*)(d), (n) * sizeof(t)))
      |                                               ^~~~~~~~~~~~~~~
/usr/perl5/5.36/lib/i86pc-solaris-thread-multi-64/CORE/sv_inline.h:468:13: note: in expansion of macro 'Zero'
  468 |             Zero(new_body, type_details->body_size, char);
      |             ^~~~
gmake: *** [Makefile:346: managesieve.o] Error 1
```

Renaming the `assert.h` header is probably also necessary for a full fix here.

--- a/imap/mailbox.c
+++ b/imap/mailbox.c
@@ -4492,7 +4492,7 @@ EXPORTED int mailbox_append_index_record(struct mailbox *mailbox,
     assert(mailbox_index_islocked(mailbox, 1));
 
     /* Append MUST be a higher UID than any we've yet seen */
-    assert(record->uid > mailbox->i.last_uid)
+    assert(record->uid > mailbox->i.last_uid);
 
     /* Append MUST have a message with data */
     assert(record->size);

--- a/imap/message.c
+++ b/imap/message.c
@@ -5070,13 +5070,13 @@ EXPORTED int message_get_priority(message_t *m, struct buf *buf)
 
 EXPORTED const struct index_record *msg_record(const message_t *m)
 {
-    assert(!message_need(m, M_RECORD))
+    assert(!message_need(m, M_RECORD));
     return &m->record;
 }
 
 EXPORTED struct mailbox *msg_mailbox(const message_t *m)
 {
-    assert(!message_need(m, M_MAILBOX))
+    assert(!message_need(m, M_MAILBOX));
     return m->mailbox;
 }
 
@@ -5096,7 +5096,7 @@ EXPORTED int message_get_size(message_t *m, uint32_t *sizep)
 
 EXPORTED uint32_t msg_size(const message_t *m)
 {
-    assert(!message_need(m, M_RECORD))
+    assert(!message_need(m, M_RECORD));
     return m->record.size;
 }
 
@@ -5110,7 +5110,7 @@ EXPORTED int message_get_uid(message_t *m, uint32_t *uidp)
 
 EXPORTED uint32_t msg_uid(const message_t *m)
 {
-    assert(!message_need(m, M_RECORD))
+    assert(!message_need(m, M_RECORD));
     return m->record.uid;
 }
 
@@ -5124,7 +5124,7 @@ EXPORTED int message_get_cid(message_t *m, conversation_id_t *cidp)
 
 EXPORTED conversation_id_t msg_cid(const message_t *m)
 {
-    assert(!message_need(m, M_RECORD))
+    assert(!message_need(m, M_RECORD));
     return m->record.cid;
 }
 
@@ -5138,7 +5138,7 @@ EXPORTED int message_get_modseq(message_t *m, modseq_t *modseqp)
 
 EXPORTED modseq_t msg_modseq(const message_t *m)
 {
-    assert(!message_need(m, M_RECORD))
+    assert(!message_need(m, M_RECORD));
     return m->record.modseq;
 }
 
@@ -5152,7 +5152,7 @@ EXPORTED int message_get_msgno(message_t *m, uint32_t *msgnop)
 
 EXPORTED uint32_t msg_msgno(const message_t *m)
 {
-    assert(!message_need(m, M_INDEX))
+    assert(!message_need(m, M_INDEX));
     return m->msgno;
 }
 
@@ -5174,7 +5174,7 @@ EXPORTED int message_get_guid(message_t *m, const struct message_guid **guidp)
 
 EXPORTED const struct message_guid *msg_guid(const message_t *m)
 {
-    assert(!message_need(m, M_RECORD))
+    assert(!message_need(m, M_RECORD));
     return &m->record.guid;
 }
 

--- a/lib/assert.h
+++ b/lib/assert.h
@@ -43,11 +43,11 @@
 #ifndef INCLUDED_ASSERT_H
 #define INCLUDED_ASSERT_H
 
-#ifdef __STDC__
-#define assert(ex)      {if (!(ex))assertionfailed(__FILE__, __LINE__, #ex);}
 void assertionfailed(const char *file, int line, const char *expr);
-#else
-#define assert(ex)      {if (!(ex))assertionfailed(__FILE__, __LINE__, (char*)0);}
-#endif
+
+#define assert(expr)                                                \
+    ((expr)                                                         \
+     ? (void)(0)                                                    \
+     : assertionfailed(__FILE__, __LINE__, #expr))
 
 #endif /* INCLUDED_ASSERT_H */

